tutorials/gcs-cf-auth/cf_auth.js

```diff
@@ -0,0 +1,133 @@
+/**
+Copyright 2018 Ian Maddox
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+'use strict';
+/* Required libraries and supporting vars */
+const path = require('path');
+process.env.NODE_CONFIG_DIR = process.env.NODE_CONFIG_DIR || path.join(__dirname, 'config');
+const config = require('config');
+const { google } = require('googleapis');
+const { Storage } = require('@google-cloud/storage');
+const validator = require('validator');
+const authServices = [];
+
+/* Project variables */
+// See config/default.json for more information
+var PROJECT_ID = config.get('general.project_id');
+const AUTH_BUCKET = config.get('auth.test_bucket');
+const DEBUG_LOGGING = config.get('logging.debug');
+const TEST_PERMISSIONS = config.get('auth.test_permissions');
+
+/**
+ * Authenticate against the GCS API. If we already have clients  instantiated for
+ * the given auth_token, return them instead of creating new ones.
+ *
+ * @param {string} authToken - The OAuth2.0 authentication token given to the CF
+ * @return Array [ds:{Datastore}, kms:{KMS}] - A cached or newly created DS and KMS object pair
+ */
+/* jshint ignore:start */
+async function authenticateAndBuildServices (authToken) {
+  try {
+    if (!authToken || !validator.isAscii(authToken)) {
+      let file = __filename;
+      debug('Invalid token provided to' + file);
+      return false;
+    }
+
+    debug(`authServices with token "${authToken}"`);
+    if (authServices[authToken] !== undefined) {
+      debug('Using cached auth objects');
+      return authServices[authToken];
+    }
+    debug('Generating new auth object(s)');
+    const oauth2client = new google.auth.OAuth2();
+    const tokenInfo = await oauth2client.getTokenInfo(authToken);
+
+    debug('token info:');
+    debug(tokenInfo);
+    let now = Date.now();
+    debug('exp:', tokenInfo.expiry_date, 'now:', Date.now(), 'remainder:', (tokenInfo.expiry_date - Date.now()));
+    if (!tokenInfo.expiry_date || tokenInfo.expiry_date < now) {
+      throw new Error('Token did not validate: ' + authToken);
+    }
+
+    // Create a cache of auth objects per token.
+    // This array contains a collection of service objects authenticated by the given
+    // token. By default, only GCS is used, but two more example blocks authenticating
```

Comment:
Be careful using 'GCS' as an abbreviation for 'Google Cloud Storage'. Since you are a Google employee, you can see the internal branding guidelines document:
https://cloud.google.com/guides/style/branding-guidelines

The guideline is to use 'Cloud Storage', instead.

Also see the guidelines for 'Cloud Datastore' and 'Cloud KMS'.